### DISCUSSION TITLE
Macthes some paths but why ?

### DISCUSSION THREAD

**Discussion by Sim4n6** [github/codeql#10653] [Q&A] [answered]
Please, I don't get well this query:

```codeql
from CallNode call
where call.getAChild*().(AttrNode).getName().matches("%path")
select call, "call" 
```
It seems that it searches for all call expressions, like `fct(..)` or `foo.bar(..)`.
Then, it casts to Attribute Node from their it gets a name and checks for `%path`.
I am kind of lost here. What is the role of `getAChild*()` with a star `*` please? 
and the `getName` corresponds to what? and attribute `name` maybe? why the value `path` and the regex `%`, please?

**Comment by hvitved** [comment, marked as answer]
`x.getAChild()` will give you you an immediate AST child of `x`. `x.getAChild*()` gives you the reflexive, transitive children of `x` (that is, including `x` itself). `(AttrNode).getName().matches("%path")` will match any attribute expression `foo.bar` where `bar` ends with `path`. So, in summary, you will get all calls where somewhere inside one of the arguments (or the receiver) there is an attribute expression that accesses an attribute whose name ends with `path`, e.g.
```py
call(x.foo) # no match
call(x.path) # match
call(x.foo.path) # match
call(x.mypath) # match
```
